src/minio/managers/policy_manager.py

```diff
@@ -780,45 +862,187 @@ async def _create_minio_policy(self, policy_model: PolicyModel) -> None:
             except Exception as e:
                 logger.warning(f"Failed to cleanup temporary policy file: {e}")
 
-    async def _update_minio_policy(self, policy_model: PolicyModel) -> None:
-        """Update a policy in MinIO by handling attachments properly."""
-        policy_name = policy_model.policy_name
+    def _generate_shadow_policy_name(self, original_name: str) -> str:
+        """Generate a unique shadow policy name."""
+        return f"{original_name}-shadow-{uuid.uuid4().hex[:8]}"
+
+    async def _create_shadow_policy(self, original_policy: PolicyModel) -> str:
+        """Create a shadow policy with the updated content."""
+        shadow_name = self._generate_shadow_policy_name(original_policy.policy_name)
+        shadow_policy = PolicyModel(
+            policy_name=shadow_name,
+            policy_document=original_policy.policy_document,
+        )
+
+        logger.info(f"Creating shadow policy: {shadow_name}")
+        await self._create_minio_policy(shadow_policy)
+        logger.info(f"Successfully created shadow policy: {shadow_name}")
+
+        return shadow_name
+
+    def _parse_policy_info(self, policy_name: str) -> tuple[str, PolicyType]:
+        """Parse policy name to extract target name and policy type."""
+        if self.is_user_home_policy(policy_name):
+            target_name = policy_name.replace(USER_HOME_POLICY_PREFIX, "")
+            policy_type = PolicyType.USER_HOME
+        elif self.is_user_system_policy(policy_name):
+            target_name = policy_name.replace(USER_SYSTEM_POLICY_PREFIX, "")
+            policy_type = PolicyType.USER_SYSTEM
+        elif self.is_group_policy(policy_name):
+            target_name = policy_name.replace(GROUP_POLICY_PREFIX, "")
+            policy_type = PolicyType.GROUP_HOME
+        else:
+            raise PolicyOperationError(
+                f"Unknown policy naming pattern: {policy_name}. "
+                f"Expected {USER_HOME_POLICY_PREFIX}, {USER_SYSTEM_POLICY_PREFIX}, or {GROUP_POLICY_PREFIX} prefix."
+            )
+        return target_name, policy_type
+
+    async def _execute_shadow_policy_update(
+        self, policy_model: PolicyModel, target_name: str, policy_type: PolicyType
+    ) -> str:
+        """Execute the shadow policy update workflow."""
+        # Step 1: Create shadow policy with new content
+        shadow_name = await self._create_shadow_policy(policy_model)
+
+        # Step 2: Attach shadow policy (user/group now has BOTH old and new access)
+        await self._attach_shadow_policy(shadow_name, target_name, policy_type)
 
-        # For user policies, detach from the specific user first
-        if policy_name.startswith(f"{TargetType.USER.value}-policy-"):
-            username = policy_name.replace(f"{TargetType.USER.value}-policy-", "")
+        # Step 3: Replace original policy with updated content
+        await self._replace_original_policy(policy_model, target_name, policy_type)
```

Comment:
Same question about redundant arguments

Reply:
then I will have to call self._parse_policy_info(policy_name) again inside of `_replace_original_policy`.  Appears to introduce additional redundancy.

Reply:
leaving them as is